### PR TITLE
ARTEMIS-1811 NIO Seq File should use RandomAccessFile with heap buffers

### PR DESCRIPTION
**Pull request by franz1981**
It use RandomAccessFile to allow using heap buffers without additional
copies and/or leaks of direct buffers, as performed by FileChannel JDK
implementation (see https://bugs.openjdk.java.net/browse/JDK-8147468)

### CODE REVIEW

**Comment by franz1981** [comment]
@wy96f @clebertsuconic This should solve the memory leak issue with the minimum amount of changes.
https://docs.oracle.com/javase/7/docs/api/java/io/RandomAccessFile.html#getChannel() doc shows that RandomAccessFile and FileChannel are required to always be in sync so NFS or other SAN storages cannot be unsynced due to using both to manipulate the underline file system

**Comment by franz1981** [comment]
Just for completeness that's what would change (ie no additional copies with heap buffers anymore):

- https://github.com/frohoff/jdk8u-jdk/blob/da0da73ab82ed714dc5be94acd2f0d00fbdfe2e9/src/share/native/java/io/RandomAccessFile.c#L79
- https://github.com/frohoff/jdk8u-jdk/blob/da0da73ab82ed714dc5be94acd2f0d00fbdfe2e9/src/share/native/java/io/io_util.c#L138
- https://github.com/frohoff/jdk8u-jdk/blob/da0da73ab82ed714dc5be94acd2f0d00fbdfe2e9/src/solaris/native/java/io/io_util_md.h#L71
- https://github.com/frohoff/jdk8u-jdk/blob/da0da73ab82ed714dc5be94acd2f0d00fbdfe2e9/src/solaris/native/java/io/io_util_md.c#L164

File: artemis-journal/src/main/java/org/apache/activemq/artemis/core/io/nio/NIOSequentialFile.java [hunk 1]
```diff
@@ -173,6 +173,40 @@ public int read(final ByteBuffer bytes) throws Exception {
       return read(bytes, null);
    }
 
+   //On *nix platform is the size by which RandomAccessFile write/read are performed using a
+   //stack-allocated copy of the java byte[]
```
Comment (michaelandrepearce) [inline artemis-journal/src/main/java/org/apache/activemq/artemis/core/io/nio/NIOSequentialFile.java]: And what about windows

Comment (franz1981) [reply artemis-journal/src/main/java/org/apache/activemq/artemis/core/io/nio/NIOSequentialFile.java]: This was just a POC commit to mitigate the issue , but we can discuss on the comments about it :)
Thanks to looking at it

Comment (franz1981) [reply artemis-journal/src/main/java/org/apache/activemq/artemis/core/io/nio/NIOSequentialFile.java]: I've verified that is a the same behaviour for Windows: it is in the shared JNI code by any OSs

**Comment by franz1981** [comment]
I'm going to "tune" the optimization while reading/writing RandonAccessFile in chunks of 8192, because on https://github.com/netty/netty/pull/9591#issuecomment-533904873 I've measured that the allocator isn't a bottleneck when chunk size > 8192 and assuming single threaded usage (malloc -> free on the same thread). 
ATM AMQP and OpenWire could cause big buffers allocation on native space (on JNI), so I would reduce the impact of this by tuning chunk size just to reduce the foot print over head of it.

File: artemis-journal/src/main/java/org/apache/activemq/artemis/core/io/nio/NIOSequentialFile.java [hunk 2]
```diff
@@ -160,14 +173,63 @@ public int read(final ByteBuffer bytes) throws Exception {
...
          if (channel == null) {
             throw new ActiveMQIllegalStateException("File " + this.getFileName() + " has a null channel");
          }
-         int bytesRead = channel.read(bytes);
+         final int bytesRead;
+         if (bytes.hasArray()) {
+            if (bytes.remaining() > BUF_SIZE) {
+               bytesRead = readRaf(rfile, bytes.array(), bytes.arrayOffset() + bytes.position(), bytes.remaining(), BUF_SIZE);
+            } else {
+               bytesRead = rfile.read(bytes.array(), bytes.arrayOffset() + bytes.position(), bytes.remaining());
```
Comment (wy96f) [inline artemis-journal/src/main/java/org/apache/activemq/artemis/core/io/nio/NIOSequentialFile.java]: @franz1981 just curious: why you added a else branch? `readRaf` seemingly can cover both conditions.

**Comment by wy96f** [comment]
> The overhead of file reading using `RandomAccessFile` and `byte[]` vs `FileChannel` and direct `ByteBuffer` is variable, but using the benchmark I've used on Netty [here](https://github.com/netty/netty/pull/9591#issuecomment-534002621) has given this result:
> 
> ```
> Benchmark                                  (chunkSize)    (fileName)  (fileSize)  (ioSize)  Mode  Cnt     Score     Error  Units
> ChunkedFileBenchmark.readFullyFile               32768  /tmp/raf.tmp    10485760     32768  avgt    4  2153.527 ± 715.641  us/op
> ChunkedFileBenchmark.readFullyFileWithNio        32768  /tmp/raf.tmp    10485760     32768  avgt    4  1775.999 ± 831.715  us/op
> ```
> 
> That means about a 25% perf difference while using super-fast disks and I assume to be the same on write side.

Do you mean 25% perf reduction using `RandomAccessFile` and `byte[]` compared to `FileChannel` and direct `ByteBuffer` ?

**Comment by franz1981** [comment]
@wy96f yes, more then a good reason to make most of the code to avoid using byte[] + rafs, but...it saves from mem leaks :)

File: artemis-journal/src/main/java/org/apache/activemq/artemis/core/io/nio/NIOSequentialFile.java (hunk 2 above)
Comment (franz1981) [reply artemis-journal/src/main/java/org/apache/activemq/artemis/core/io/nio/NIOSequentialFile.java]: to save inlining budget: for most of the cases the first check will safe the call on the happy path

Comment (wy96f) [reply artemis-journal/src/main/java/org/apache/activemq/artemis/core/io/nio/NIOSequentialFile.java]: Understand :)

**Pull request closed without merging**